lib/active_merchant/billing/gateways/quickbooks.rb

```diff
@@ -0,0 +1,312 @@
+      end
+
+      def add_payment(post, payment, options = {})
+        add_creditcard(post, payment, options)
+      end
+
+      def add_creditcard(post, creditcard, options = {})
+        card = {}
+        card[:number] = creditcard.number
+        card[:expMonth] = "%02d" % creditcard.month
+        card[:expYear] = creditcard.year
+        card[:cvc] = creditcard.verification_value if creditcard.verification_value?
+        card[:name] = creditcard.name if creditcard.name
+        card[:commercialCardCode] = options[:card_code] if options[:card_code]
+
+        post[:card] = card
+      end
+
+      def add_context(post, context)
+        payment_context = {}
+        payment_context[:tax] = context[:tax] if context[:tax]
+        payment_context[:recurring] = context[:recurring] if context[:recurring]
+
+        post[:context] = payment_context
+      end
+
+      def auth_object_from(authorization)
+        uri = "#{ENDPOINT}/#{authorization}"
+        commit(uri, {}, :get)
+      end
+
+      def parse(body)
+        JSON.parse(body)
+      end
+
+      def commit(uri, body = {}, method = :post)
+        endpoint = gateway_url + uri
+        begin
+          response = case method
+          when :post
+            ssl_post(endpoint, post_data(body), headers(method: :post, uri: endpoint))
+          when :get
+            ssl_request(:get, endpoint, nil, headers(method: :get, uri: endpoint))
+          end
+        rescue ResponseError => e
+          if e.response.body.present?
+            response = e.response.body
+          else
+            return Response.new(
+              false,
+              "There was a problem connecting to QuickBooks",
+              {
+                code: e.response.code,
+                message: e.response.message
+              },
+              test: test?
+              )
+          end
+        end
+
+        response_object(response)
+      end
+
+      def response_object(raw_response)
+        parsed_response = parse(raw_response)
+
+        Response.new(
+          success?(parsed_response),
+          message_from(parsed_response),
+          parsed_response,
+          authorization: authorization_from(parsed_response),
+          test: test?,
+          cvv_result: cvv_code_from(parsed_response),
+          error_code: errors_from(parsed_response)
+        )
+      end
+
+      def gateway_url
+        test? ? test_url : live_url
+      end
+
+      def post_data(data = {})
+        data.to_json
+      end
+
+      def headers(options)
+        return unless options[:method] && options[:uri]
+
+        method = options[:method]
+        uri = options[:uri]
+        request_uri = URI.parse(uri)
+
+        request_class = case method
+          when :post
+            Net::HTTP::Post
+          when :get
+           Net::HTTP::Get
+        end
+
+        # required for Ruby 1.9 backwards compatibilty
+        begin
+          request_object = request_class.new(request_uri)
+        rescue
```

Comment:
I think we're safe here since we're in a private helper method. The exceptions that flare up are confusing, so to keep things clean, I'd prefer to keep this the way it is -- since this is just about constructing a request object appropriately for the Ruby being used. The Oauth `headers` method will barf if the URI is malformed anyway.